clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -453,7 +453,8 @@ public void testSubscription(GroupProtocol groupProtocol) {
 
         consumer.assign(singletonList(tp0));
         assertTrue(consumer.subscription().isEmpty());
-        assertEquals(singleton(tp0), consumer.assignment());
+        TestUtils.waitForCondition(() -> consumer.assignment().equals(singleton(tp0)),
+                "Expected consumer to be assigned the partition");
```

Comment:
If the suggestion on my [comment](https://github.com/apache/kafka/pull/16673/files#r1697100482) above works, I would expect we won't need to change this test like this (shows a rather undesired behaviour of the assign I would say). We would end with the test unchanged probably, with the assignment effectively changed when assign() returns for both consumers.